### PR TITLE
Update legacy structs during compile

### LINKED ISSUE #3058: Reflection stripping for shader lib target appears to be broken/regressed from 2019 version
Observe this simple shader, compiling with /Qstrip_reflect works for the compute shader target but not as a library target. Interestingly it's not specific to raygen, as I can select export for compute shader as well.

This appears to be a regression from 2019 release which strips reflection data as expected.

Repro:

dxc /auto-binding-space 1 /Qstrip_debug /Qstrip_reflect /Qstrip_private /E ShaderCompute /Zpr /O3 /T cs_6_3 /FcShaderReflectionTest.compute.d3dasm ShaderReflectionTest.shader

dxc /auto-binding-space 1 /Qstrip_debug /Qstrip_reflect /Qstrip_private /exports ShaderRGS /Zpr /O3 /T lib_6_3 /FcShaderReflectionTest.raygen.d3dasm ShaderReflectionTest.shader

Shader:

Texture2D DummyTexture;

[shader("compute")]
[numthreads(1, 1, 1)] 
void ShaderCompute()
{
	return;
}

[shader("raygeneration")] void ShaderRGS()
{
	return;
}

Attached output generated: 

[ShaderReflectionTest.compute.d3dasm.txt](https://github.com/microsoft/DirectXShaderCompiler/files/5003779/ShaderReflectionTest.compute.d3dasm.txt)
[ShaderReflectionTest.raygen.d3dasm.txt](https://github.com/microsoft/DirectXShaderCompiler/files/5003780/ShaderReflectionTest.raygen.d3dasm.txt)

### PR DESCRIPTION
**Pull request by pow2clk**
Instead of bailing out when compiling libraries without full 16-bit
support enabled, we update the struct type for legacy layouts and
replace the associated ops to the new types where necessary.

This results in converted types named with a leading "dx." prefix that
is an array type which often isn't expected in validation where it newly
shows up, so a change there to permit these types to pass.

As an incidental, remove function annotations when appropriate
regardless of full 16-bit enabling. There was no reason to make this
depend on that setting.

Adapt tests to the new types. Library targets may now produce a
dx.alignment* type instead of the user-specified type.

Fixes #3058

### CODE REVIEW

File: lib/HLSL/DxilCondenseResources.cpp [hunk 1]
```diff
@@ -553,9 +553,16 @@ class DxilLowerCreateHandleForLib : public ModulePass {
...
 
-    if (m_bIsLib || m_bLegalizationFailed)
+    if (m_bLegalizationFailed)
       return bChanged;
 
+    if (m_bIsLib) {
+      if (DM.GetOP()->UseMinPrecision())
+        UpdateStructTypeForLegacyLayout();
+
+      return true;
```
Comment (tex3d) [inline lib/HLSL/DxilCondenseResources.cpp:563]: Why not return `bChanged` here?

File: lib/HLSL/DxilContainerReflection.cpp [hunk 2]
```diff
@@ -1085,7 +1085,7 @@ HRESULT CShaderReflectionType::Initialize(
 
       // Try to "clean" the type name for use in reflection data
       llvm::StringRef name = structType->getName();
-      name = name.ltrim("dx.alignment.legacy.");
+      name = name.ltrim(kHostLayoutTypePrefix);
```
Comment (tex3d) [inline lib/HLSL/DxilContainerReflection.cpp:1088]: Should this keep both old and new trim in case IR is from prior version of compiler?

**Review by tex3d — APPROVED** [review]
Looks good with a couple minor suggestions.

File: lib/HLSL/DxilCondenseResources.cpp (hunk 1 above)
Comment (pow2clk) [reply lib/HLSL/DxilCondenseResources.cpp:563]: It's been almost two months and I don't remember. I made the change at the same time that I changed the prefix ( d5ada27a2a ). 

I don't think just returning `bChanged` is right. If the type names have changed, I think we need to return true. `UpdateStructTypeforLegacyLayout` doesn't currently indicate if it changed anything or not. I think I'll make it tell what it did and return `bChanged || bLocalChanged` where `bLocalChanged` is the result.

Does that sound right?

Comment (pow2clk) [reply lib/HLSL/DxilCondenseResources.cpp]: This sounds familiar. So that's what I've done.

File: lib/HLSL/DxilContainerReflection.cpp (hunk 2 above)
Comment (pow2clk) [reply lib/HLSL/DxilContainerReflection.cpp:1089]: done

**Review by pow2clk — COMMENTED** [review]
respond to comments

File: lib/HLSL/DxilCondenseResources.cpp (hunk 1 above)
Comment (tex3d) [reply lib/HLSL/DxilCondenseResources.cpp]: Yes, your change is good.

**Pull request merged**